### ISSUE TITLE
Some bugs in "get_schedule" API

### ISSUE THREAD

**Issue by iambocai** [jhuckaby/Cronicle#10]
Conicle's "get_schedule"  API output sometimes does not meet expectations, sometimes incomplete, I test it both by curl(Linux) and  Chrome(Windows), it seems the bug is more likely came from the server side.

> OS: CentOS 6.5 x64
> Node: 7.0.0
> Cronicle: 0.6.5

1. **[RIGHT]** Wants 2 returns 2:
![image](https://cloud.githubusercontent.com/assets/2259715/22962570/9469f5f2-f388-11e6-8d2d-8512d7228155.png)
(note: 8012 is the cronicle server port)

2. **[WRONG]** Want 2 but returns 16:
![image](https://cloud.githubusercontent.com/assets/2259715/22962294/01dd75c0-f387-11e6-9f45-022ab945aab8.png)

3. **[WRONG]** Truncated output:
(/api/app/get_schedule/v1?offset=16&limit=2)
![image](https://cloud.githubusercontent.com/assets/2259715/22962507/1df40cd2-f388-11e6-9dba-682f2d6b2b14.png)

**Comment by jhuckaby** [comment]
Hmm, this is concerning:

<img width="208" alt="screenshot 2017-02-14 22 47 46" src="https://cloud.githubusercontent.com/assets/700765/22963371/a54cbb42-f307-11e6-91b9-bfbc63e6fdf6.png">

Your schedule is on page -17, with 50 items per page, so you've been "churning" through the items somehow.  Like, adding new ones and deleting old ones.  You only have 116 items currently, but you've churned through ~850 of them, in order to move the needle all the way to page -17.

Do you have some kind of script that creates and deletes items from the schedule automatically?  I think there may be some kind of race condition here, and your script sent in multiple requests at the same time, and the list data got corrupted on disk.  Cronicle does have in-memory locks, but there are various ways to defeat them.

I will try to reproduce your use case by artificially expanding my schedule so I get to page -17, ending up with 116 items, and see if I can get any API pagination errors.

I know you obviously have some private company data in your schedule, but what I really need to see is the state of the list data on disk, so I can tell if there is any corruption.  If you could, please run this command and copy & paste the output here.  It should not contain any of your private information, just some metadata about the data pagination system:

```
cd /opt/cronicle
./bin/storage-cli.js list_info global/schedule
```

Thanks.

**Comment by iambocai** [comment]
> Do you have some kind of script that creates and deletes items from the schedule automatically?

No, “get_schedule” API is the first API I tried to use ( to backup schedule configs ), all the tasks were added by my hand,   :)  

And here is the output:
> [root@host cronicle]# ./bin/storage-cli.js list_info global/schedule
> List Header: global/schedule: {
>         "page_size": 50,
>         "first_page": -17,
>         "last_page": -15,
>         "length": 116,
>         "type": "list"
> }
> 
> Page -17: 50 items
> Page -16: 50 items
> Page -15: 16 items

I was also confused about the first_page last_page number，why are they all negative integers， what's the right way for me to get all schedules ？

**Comment by iambocai** [comment]
> [root@host cronicle]# curl -H 'X-API-Key: 225f7c95648a6909f765733b********' 'http://localhost:8012/api/app/get_schedule/v1?offset=-850&limit=10'
> {"code":0,"rows":[null,null,null,null,null,null,null,null,null,null],"list":{"page_size":50,"first_page":-17,"last_page":-15,"length":116,"type":"list"}}
> [root@host cronicle]# curl -H 'X-API-Key: 225f7c95648a6909f765733b********' 'http://localhost:8012/api/app/get_schedule/v1?offset=850&limit=10' 
> {"code":0,"rows":[],"list":{"length":0}}
> [root@host cronicle]# curl -H 'X-API-Key: 225f7c95648a6909f765733b********' 'http://localhost:8012/api/app/get_schedule/v1?offset=-750&limit=10'
> {"code":0,"rows":[null,null,null,null,null,null,null,null,null,null],"list":{"page_size":50,"first_page":-17,"last_page":-15,"length":116,"type":"list"}}

**Comment by jhuckaby** [comment]
Thank you for the detailed information.  I see at least two different bugs here, and I will fix them as soon as possible.

> I was also confused about the first_page last_page number，why are they all negative integers
> schedule offset seems start at 0, not -850:

Yes, sorry for the confusion over this.  The page numbers are just an internal representation of a paginated linked list.  The are negative simply because new schedule entries are added to the "front" of the list, not the "back", and so the pagination system adds new pages below zero.  However, this is all internal stuff, and you should not have to worry about it, or ever see it.  The publicly facing `offset` always starts from zero and goes up (positive).  The internal back-end API then maps this to its page system, which should be invisible to the user.

I was just gathering information so I could reproduce the bug reliably on my side, and get my storage system and linked list pages into the exact same state as yours.

You should never have to worry about any of this, but if you are interested in learning more about the internal storage system, it's fully documented here: [pixl-server-storage](https://github.com/jhuckaby/pixl-server-storage), specifically the [Lists](https://github.com/jhuckaby/pixl-server-storage#lists) section.

> what's the right way for me to get all schedules ？

You are already doing everything right.  There is simply a bug (well, more like two bugs) preventing the API from working properly.  I will fix this ASAP.  Once everything is working again, if you want to fetch the *entire* schedule, just make sure the `offset` is set to 0, and the `limit` is set to a large number like `9999`.  Like this:

```
GET /api/app/get_schedule/v1?offset=0&limit=9999
```

I'm also going to change the `get_schedule` API right now, so if you omit the `limit` parameter, it will default to the entire list length (right now it defaults to `50` which makes no sense).

Thanks again for this bug report.  I'm working on fixing all this stuff right now.

**Comment by jhuckaby** [comment]
Okay, everything should be fixed now, in Cronicle v0.6.7: https://github.com/jhuckaby/Cronicle/releases/tag/v0.6.7

Note that the API truncated output bug was in a dependency library, fixed here:
https://github.com/jhuckaby/pixl-server-web/issues/2

So if you use the Cronicle upgrade script, it should update all the dependencies as well:

```
/opt/cronicle/bin/control.sh upgrade
```

Thanks for this bug report!

\- Joe

**Comment by iambocai** [comment]
wow~  I just upgrade my server and it works fine,  thank you very much! 🥇 💯
